### PR TITLE
curvefs/mds: Change the policy of create copyset

### LINKED ISSUE #841: curvefs: improve the policy of create copyset
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**
The policy of create copyset is random now. The randomly choose metaserver maybe unbalanced.

**Describe the solution you'd like (描述你期望的解决方法)**
Change the policy of create copyset.

**Describe alternatives you've considered (描述你想到的折衷方案)**
Use the policy based on disk capacity.

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by SeanHai**
The randomly choose metaserver maybe unbalanced.
Use the policy based on disk capacity.

### What problem does this PR solve?

Issue Number: close #841  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/mds/topology/topology.h [hunk 1]
```diff
@@ -200,23 +201,24 @@ class Topology {
...
-    virtual TopoStatusCode ChooseSinglePoolRandom(
-        PoolIdType *out,
-        const std::set<PoolIdType> &unavailablePools) const = 0;
-
     virtual TopoStatusCode ChooseZonesInPool(
         PoolIdType poolId, std::set<ZoneIdType> *zones,
         const std::set<ZoneIdType> &unavailableZones, int count) const = 0;
 
     virtual TopoStatusCode ChooseSingleMetaServerInZone(
         ZoneIdType zoneId, MetaServerIdType *metaServerId,
```
Comment (cw123) [inline curvefs/src/mds/topology/topology.h:215]: replace these in recover schedule, and delete these code.

File: curvefs/src/mds/topology/topology.cpp [hunk 2]
```diff
@@ -1257,6 +1235,88 @@ TopoStatusCode TopologyImpl::ChooseSingleMetaServerInZone(
...
 }
 
+TopoStatusCode TopologyImpl::ChooseAvailableMetaServers(
+    std::set<MetaServerIdType> *metaServers,
+    PoolIdType *poolId) {
+    ReadLockGuard rlockMetaserver(metaServerMutex_);
+    std::vector<MetaServer> vec;
+    for (const auto &it : metaServerMap_) {
+        vec.emplace_back(it.second);
+    }
```
Comment (wu-hanqing) [inline curvefs/src/mds/topology/topology.cpp:1245]: it seems no need to copy all metaserver from map to vector, directly store metaserver's pointer is meeting current purpose.

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/src/mds/topology/topology.h (hunk 1 above)
Comment (SeanHai) [reply curvefs/src/mds/topology/topology.h]: > replace these in recover schedule, and delete these code.

fixed

File: curvefs/src/mds/topology/topology.cpp (hunk 2 above)
Comment (SeanHai) [reply curvefs/src/mds/topology/topology.cpp]: > it seems no need to copy all metaserver from map to vector, directly store metaserver's pointer is meeting current purpose.

fixed

File: curvefs/src/mds/schedule/topoAdapter.cpp [hunk 3]
```diff
@@ -298,34 +298,19 @@ bool TopoAdapterImpl::CreateCopySetAtMetaServer(CopySetKey id,
...
+    MetaServerIdType *target) {
+    TopoStatusCode ret = topo_->ChooseRecoveredMetaServer(poolId, excludeZones,
+        excludeMetaservers, target);
+    if (ret == TopoStatusCode::TOPO_OK) {
+        return true;
     }
-
-    *zoneId = *zones.begin();
-    return true;
+    return false;
```
Comment (wu-hanqing) [inline curvefs/src/mds/schedule/topoAdapter.cpp:311]: just `return ret == TopoStatusCode::TOPO_OK;`

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/src/mds/schedule/topoAdapter.cpp (hunk 3 above)
Comment (SeanHai) [reply curvefs/src/mds/schedule/topoAdapter.cpp:311]: > just `return ret == TopoStatusCode::TOPO_OK;`

fixed.

File: curvefs/src/mds/topology/topology.cpp [hunk 4]
```diff
@@ -1129,129 +1133,118 @@ TopoStatusCode TopologyImpl::UpdatePartitionTxIds(
...
+                } else {
+                    LOG(ERROR) << "get pool failed when choose metaservers,"
+                               << " the poolId = " << it.first;
+                    return TopoStatusCode::TOPO_POOL_NOT_FOUND;
+                }
+            } else {
+                replicaNum = iter->second;
+            }
+            // find the available metaserver group
+            if (it.second.size() == replicaNum) {
```
Comment (cw123) [inline curvefs/src/mds/topology/topology.cpp:1241]: what if if (it.second.size() >= replicaNum)

File: curvefs/src/mds/topology/topology.cpp [hunk 5]
```diff
@@ -1129,129 +1133,118 @@ TopoStatusCode TopologyImpl::UpdatePartitionTxIds(
...
-    poolMutex_.RDLock();
-    std::unordered_map<PoolIdType, Pool> tmpPoolMap = poolMap_;
-    poolMutex_.Unlock();
-
-    // erase unavailable pools
-    for (const auto &pool : unavailablePools) {
-        auto it = tmpPoolMap.find(pool);
-        if (it != tmpPoolMap.end()) {
-            tmpPoolMap.erase(it);
+TopoStatusCode TopologyImpl::ChooseRecoveredMetaServer(
```
Comment (cw123) [inline curvefs/src/mds/topology/topology.cpp:1136]: The algorithm should be similar to ChooseAvailableMetaServers, considering disk capacity

File: curvefs/src/mds/topology/topology.cpp (hunk 4 above)
Comment (SeanHai) [reply curvefs/src/mds/topology/topology.cpp:1241]: > what if if (it.second.size() >= replicaNum)

I think it makes no difference.

File: curvefs/src/mds/topology/topology.cpp (hunk 5 above)
Comment (SeanHai) [reply curvefs/src/mds/topology/topology.cpp:1136]: > The algorithm should be similar to ChooseAvailableMetaServers, considering disk capacity

fixed

**Review by cw123 — APPROVED** [review]

**Pull request merged**